examples/step-93/doc/intro.dox

```diff
@@ -0,0 +1,369 @@
+for the programer to know about) rather than enforcing it
+explicitly (say, in the form of a function `mesh_dof_to_linear_system_index()`).
+It is then very easy for a number of bugs to emerge.
+For example, we can not renumber all degrees of freedom with the
+functions in DoFRenumbering in arbitrary ways if we want the $m$
+additional variables always at the end.
+
+Precisely because it is so easy to make mistakes and because it is so
+restricting to have implicit conventions like "Yes, your matrix can be
+larger than the number of DoFs in the DoFHandler; but it can't be
+smaller", deal.II a long time ago made the decision that a lot of
+functions test that the sizes of vectors and matrices are equal to the
+number of unknowns in a DoFHandler. For example,
+DoFTools::make_sparsity_pattern() takes a DoFHandler and expects that
+the sparsity pattern has as many rows and columns as there are DoFs in
+the DoFHandler. So does
+DoFTools::make_hanging_node_constraints(). VectorTools::interpolate_boundary_values()
+uses the same convention. Many many other functions do too, and they
+all in fact check that this is true and will terminate the program if
+it is not. This is because one can find many many bugs in codes this
+way, and it also allows for an assumption that is so obvious that we
+have never stated it so far: The degree of freedom with index $i$
+corresponds to row $i$ of the linear system and column $i$ of the
+matrix. If we allowed sizes to differ, we would need to have a way to
+make this connection. Further, that way would need to be be explicit (by way of a
+function that translates from DoF index to row and column, as already mentioned above) or there
+would be no end to the possibility of introducing bugs.
+
+In other words, the fact that we want to have $2n$ degrees
+of freedom in the DoFHandler, but matrices and vectors with
+$2n + m$ rows and columns does not work with deal.II's
+underlying design decisions. Creating matrices and vectors that happen
+to be larger is going to trigger lots of assertions, and for good
+reasons.
+
+
+
+<h3>How do we address this?</h3>
+
+So then how are we going to address this? At its core, what needs to
+happen is that the DoFHandler knows about the additional degrees of
+freedom. The question is how we are going to teach it about these
+"nonlocal" degrees of freedom -- they are, after all, not ones that
+are associated with vertices, edges, faces, or cells of the mesh as
+the usual finite element degrees of freedom are. The approach we are
+going to use is not what one would generally call "obvious" or
+"natural", but it does fit into the infrastructure deal.II provides
+(even though this infrastructure was originally built for other
+purposes). In fact, at the time this program is written in 2024, there
+is substantial debate on whether we can implement *better* ways to achieve
+the same result that do not require this kind of "abuse" of features intended
```

Comment:
We should mention at some point that using `DoFHandler` in hp-mode is more expensive than non-hp mode, so there is a runtime cost to this approach.

Reply:
Perhaps true, but do you think it makes a difference? In my timings the DoFHandler operations never seem to show up very high on profiles. There is, of course, also the issue that we don't currently have an alternative and it seems not useful to point out that A is more expensive than B if A is your only option.

Reply:
That's a fair point - lets not add it in here.